js/bot.js

```diff
@@ -1,168 +1,230 @@
-      })
-      const followersWhoWeShouldFollowPromise = removeUsersWhoShouldntBeSentAFollow(
-        followersWhoHaventBeenFollowedIds
-      )
-      const followNewUsersPromises = followersWhoWeShouldFollowPromise.then(followersWhoWeShouldFollowIds => {
-        return Promise.all(
-          followersWhoWeShouldFollowIds.map(followerId => {
-            return followUser(followerId)
+function compareFollowersToFollowing() {
+  console.info("Handling new and old followers");
+  return getAccountId().then((accountId) => {
+    return getFollowersAndFollowing(accountId).then(
+      ({ followerIds, followingIds }) => {
+        // follow users that are following the bot
+        // and also havent already been followed or followe request before
+        const followersWhoHaventBeenFollowedIds = followerIds.filter(
+          (followerId) => {
+            const isFollowingUser = followingIds.includes(followerId);
+            return !isFollowingUser;
+          }
+        );
+        const followersWhoWeShouldFollowPromise =
+          removeUsersWhoShouldntBeSentAFollow(
+            followersWhoHaventBeenFollowedIds
+          );
+        const followNewUsersPromises = followersWhoWeShouldFollowPromise.then(
+          (followersWhoWeShouldFollowIds) => {
+            return Promise.all(
+              followersWhoWeShouldFollowIds.map((followerId) => {
+                return followUser(followerId);
+              })
+            );
+          }
+        );
+
+        // unfollow users the bot follows that arent following the bot
+        const followingWhoHaveUnfollowedIds = followingIds.filter(
+          (followingId) => {
+            const isFollowedBackByUser = followerIds.includes(followingId);
+            return !isFollowedBackByUser;
+          }
+        );
+        const unfollowOldUsersPromises = Promise.all(
+          followingWhoHaveUnfollowedIds.map((followingId) => {
+            return unfollowUser(followingId);
           })
-        )
+        );
+
+        return Promise.all([
+          followNewUsersPromises,
+          unfollowOldUsersPromises,
+        ]).then((results) => {
+          const [followedUsers, unfollowedUsers] = results;
+          return { followedUsers, unfollowedUsers };
+        });
+      }
+    );
+  });
+}
+
+function processNotificationEvent(message) {
+  const { status, account, type } = message.data;
+
+  // if a user follows the bot
+  if (type === "follow") {
+    followUser(account.id)
+      .then((result) => {
+        console.info("Followed back: ", result);
       })
+      .catch(console.error);
+  }
+
+  // if a user favourites the bot's toot
+  if (type === "favourite") {
+    // check if it's a deletable toot
+    if (!status.content.includes(FAVOURITE_TOOT_TO_DELETE_STRING)) {
+      return;
+    }
+
+    /*
+      check if the account that favourited owned
+      the toot that our bot had replied to
+    */
+    if (account.id === status.in_reply_to_account_id) {
```

Comment:
This is failing for my case. The bot replied to something I boosted from another account.